### PR TITLE
Adds some geography. This one is a kind of big deal.

### PR DESCRIPTION
**Pull request by I-am-Erk**
#### Summary
Features "The map changes as you travel east-west and north-south, revealing either rural areas or oceans and megacities"

#### Purpose of change
Our current mapgen creates a sort of weird suburban zone. It's fine but it's bland. There's very little reason to explore outside one overmap. There are no real cities. No ocean. No mountains. No vast forests.

#### Describe the solution
Adds these things.

1. As you travel West, the cities don't get smaller but they start to space out more.
2. As you travel East, the cities get bigger and more crowded until you're basically in a constant megacity.
3. As you travel West and North, the forests get denser and thicker until you're soon in a rural forest interspersed by roads and campgrounds
4. In the eastern megacity as you travel north, the cities remain big but get less dense again.

#### Describe alternatives you've considered
The massachussets mod would be another way to do this but I've wanted to boost the mapgen abilities for a while, and it turned out to be easier than I thought.

I wanted to add oceans in this PR (see screenshot below) but they were causing things to hang on generation and I don't want to introduce too many points of failure in one go.

#### Testing
here's a large image showing just the change to cities:
image

This shows forest density increasing as you head West from the starting map
image

This is a city at overmap 10,10, which is getting into the outskirts of boston.
image

When you get into the really rural areas approaching what would be vermont, you get... rural areas.
image

But zooming in, those areas still have roads and points of interest:
image-1

This is just a cool shot around the 5,-2 area:
image
I can't get over how much cooler this is because here's a shot from in the same worldgen, a bit more to the west and north:
image

#### Additional context
this one shows the later ocean change. It's super cool but I have moved it to a later PR so I can push the stuff that's working right... the ocean stuff causes worldgen to hang and crash, and the fix is not super simple.
image

### CODE REVIEW

**Comment by Qrox** [comment]
Do you plan to jsonify this? I think the math parser for EOC could be used for it.

**Comment by Terrorforge** [comment]
Will this bias scenarios like Missed towards dumping you in the eastern megacities? Or is that not how starting locations work

Conversely, would it be ossible to do that on purpose? It would be fun to have scenarios where you start right on the coast in the biggest megacities or far to the west in the most rural areas.

**Comment by I-am-Erk** [comment]
Qrox: I do definitely plan to expose it to JSON, but not this PR. For now I just want it working and getting bug tested.

Terrorforge: I don't think scenarios will usually search far enough afield to find the big cities at this point, but I would like some infrastructure and logic related to them so that we can have scenarios where you start in the city or on an island in the ocean at some point. There's nothing preventing it

File: src/overmap.cpp [hunk 1]
```diff
@@ -5184,6 +5216,7 @@ void overmap::place_cities()
             }
             // Ensure that cities are at least size 2, as city of size 1 is just a crossroad with no buildings at all
             size = std::max( size, 2 );
+            size = std::min( size, 55 );
```
Comment (ehughsbaird) [inline src/overmap.cpp:5219]: Why this restriction?

File: src/overmap.cpp [hunk 2]
```diff
@@ -4500,6 +4500,15 @@ void overmap::place_forest_trailheads()
...
     const oter_id default_oter_id( settings->default_oter[OVERMAP_DEPTH] );
+    const point_abs_om this_om = pos();
+    float forest_size_scalar = 0.0f;
+    if( this_om.x() < 0 ) {
+        forest_size_scalar += static_cast<float>( this_om.x() * -0.02f );
+    }
+    if( this_om.y() < 0 ) {
+        forest_size_scalar *= static_cast<float>( 1.0f + this_om.y() * -0.1f );
+    }
+    forest_size_scalar = std::min( forest_size_scalar, 0.5f );
```
Comment (ehughsbaird) [inline src/overmap.cpp:4511]: What does this restriction do? Why 0.5?

File: src/overmap.cpp [hunk 3]
```diff
@@ -5115,11 +5124,32 @@ spawns happen at... <cue Clue music>
 20:56 <kevingranade>: game:pawn_mon() in game.cpp:7380*/
 void overmap::place_cities()
 {
+    int city_size_scalar = 0; // used to increase city size as we move East and South.
+    int city_space_scalar =
+        0; // used to space cities out as we go more into the West and the Appalachians.
```
Comment (ehughsbaird) [inline src/overmap.cpp:5129]: This looks a little better.
```suggestion
    // used to space cities out as we go more into the West and the Appalachians.
    int city_space_scalar = 0;
```

File: src/overmap.cpp [hunk 4]
```diff
@@ -5115,11 +5124,32 @@ spawns happen at... <cue Clue music>
 20:56 <kevingranade>: game:pawn_mon() in game.cpp:7380*/
 void overmap::place_cities()
 {
+    int city_size_scalar = 0; // used to increase city size as we move East and South.
+    int city_space_scalar =
```
Comment (ehughsbaird) [inline src/overmap.cpp:5128]: These aren't really scalars. You're adding them, not multiplying by them.

File: src/overmap.cpp [hunk 5]
```diff
@@ -4500,6 +4500,15 @@ void overmap::place_forest_trailheads()
 void overmap::place_forests()
 {
     const oter_id default_oter_id( settings->default_oter[OVERMAP_DEPTH] );
+    const point_abs_om this_om = pos();
+    float forest_size_scalar = 0.0f;
```
Comment (ehughsbaird) [inline src/overmap.cpp:4504]: This is also not a scalar.

File: src/overmap.cpp (hunk 1 above)
Comment (I-am-Erk) [reply src/overmap.cpp:5219]: If a city gets too large it's impossible to place.

**Comment by fairyarmadillo** [comment]
Aren't the southeastern parts of New England swampier? Could do something with that.

Looking great btw!

**Comment by OtherBiscuit** [comment]
Oh yeah, this is good. Personally, I'd relegate it to a mod for those without processing capability for this, but it's really cool!

**Comment by I-am-Erk** [comment]
It doesn't really add any processing. The cities are no more intensive than any other city.

**Comment by I-am-Erk** [comment]
It doesn't presently come close to any capacity for that. It's possible to do that and not really difficult but it's entirely separate.

Basically this change causes gradual shifts but doesnt make any strict assertions saying "it's city time now" or anything. We could mix this up with some actual biome changes covered in regional settings, and that might actually be a really interesting way to do some stuff.

**Comment by DeftGrain** [comment]
Absolutely, just looking at the generation itself this has a lot of potential and opens up a lot of possibilities. The map generation itself reminds me a bit of UnReal World mapgen in a very good way. You have the general expectation of something in a particular direction, but it's still procedurally done. I feel like this will open up the opportunity to start work on a framework for more realistic storms and temperature drops during winter, adding another layer to the game, especially with the dicussed addition of tile slickness values for winter. An excellent addition Erk.

**Comment by Inglonias** [comment]
A little late, but I think this is super neat. Fantastic work.

**Comment by ashGlaw** [comment]
Erk, thank you. This is wonderful! I've wanted this for a long time now. Bless

**Comment by Irets** [comment]
This looks very cool - can't wait for my compiler to put together a fresh copy of the game.
You didn't mention anything about what's expected at south. How do these changes apply to travelling south?

**Comment by Inglonias** [comment]
It occurs to me that while this is super cool, we should probably find a way to tell people about this. If people go from 0.G to 0.H and mapgen is sufficiently different in long-running games, people might get confused.

**Comment by I-am-Erk** [comment]
> It occurs to me that while this is super cool, we should probably find a way to tell people about this. If people go from 0.G to 0.H and mapgen is sufficiently different in long-running games, people might get confused.

It shouldn't have that big an effect that it needs to be anywhere besides the changelog. THe worst thing someone might see is a sharp, ugly line at the edge of the changed mapgen where the gradient jumps suddenly. Nothing ought to break.

**Pull request merged**